### ISSUE TITLE
Feature request: add stopping condition to fmin

### ISSUE THREAD

**Issue by faye36** [hyperopt/hyperopt#449]
It would be useful to add a way to stop the hyperparameter search early after evaluating one or more trials  where the loss has fallen below a certain threshold - instead of always continuing the search until max_evals is reached.

**Comment by ninnghazad** [comment]
the way i do that is to loop fmin with max_evals=len(trials)+batch_size and check conditions after each batch. batch_size i usually choose by rule-of-thumb * number-of-workers or something like that.

**Comment by PhilipMay** [comment]
What do you think about a timed stopping condition? Instead of telling hyperopt to run for a certain number of evaluations you could tell hyperapt to run x minutes?

**Comment by ninnghazad** [comment]
that'd be a nice addition - can't cleanly do that from the outside. however, the loss-based option also could have the advantage over external solutions that (i think) the state of the tpe would not be interrupted to check on the stopping condition.

**Comment by PhilipMay** [comment]
What about a time based early stopping? If loss does not improve after x minutes then leave it. Or a mix of time and iterations...

**Comment by maxpumperla** [comment]
I think the loss-based stopping is more useful in practice, but why not give users options? We could think about a simple `stopping_criterion` argument to fmin that is None by default and takes a function that evaluates as boolean to tell fmin when to stop. We can provide a few default implementations, like time- or loss-based.

**Comment by PhilipMay** [comment]
> I think the loss-based stopping is more useful in practice, but why not give users options? We could think about a simple `stopping_criterion` argument to fmin that is None by default and takes a function that evaluates as boolean to tell fmin when to stop. We can provide a few default implementations, like time- or loss-based.

Great suggestion. Will have a look into it.

**Comment by maxpumperla** [comment]
@PhilipMay just a rough idea, may not even work as suggested. but if you come up with a good idea, I'm happy to discuss, as you know.

**Comment by jasonminsookim** [comment]
> the way i do that is to loop fmin with max_evals=len(trials)+batch_size and check conditions after each batch. batch_size i usually choose by rule-of-thumb * number-of-workers or something like that.

Could you provide example code for this?

**Comment by ninnghazad** [comment]
@jasonminsookim something like this. it processes trials in chunks, args.trials can be a filename (serial execution) or a mongo-url (parallel execution):
```
                if args.trials.startswith("mongo://"):
			trials = MongoTrials(args.trials, exp_key=args.exp_key)
			mongomode = True
		else:
			mongomode = False
			try:
				print("Trying to load trials from '{}' ...".format(args.trials))
				trials = pickle.load(open(args.trials, "rb"))
				print("Found saved Trials! Loading...")
				trials_step = 1
				print("Rerunning: {} finished trials, will run {} more trials".format(
					len(trials.trials), args.max_trials))
			except:  # create a new trials object and start searching
				print("Starting new trials:", args.trials)
				trials = Trials()

		for i in range(args.max_trials):
			fmin(
				fn=optimizer.f,
				space=ns.space,
				algo=tpe.suggest,
				trials=trials,
				max_evals=len(trials) + 64, # this limits the amount of (mongo)workers doing things in parallel
				max_queue_len=32,
				verbose=args.verbose
			)

			print("Currently finished:",len(trials))
			print("Average best loss:",trials.average_best_error())
			print("Current best loss:",trials.best_trial['result']['loss'])

	        	best = trials.best_trial['result']['space']
			pp.pprint(best)
	
                        # check for abort crit somewhere around here

			filename = 'models/params_'+ str(trials.best_trial['result']['loss']) + "_" + \
				datetime.datetime.now().strftime("%Y-%m-%d_%H:%M:%S") + '.pickle'
			with open(filename, 'wb') as output:
				pickle.dump(best,
							output, pickle.HIGHEST_PROTOCOL)
			print('saved as', filename)
			# save the trials object, maybe
			if not mongomode:
				with open(args.trials, "wb") as f:
					pickle.dump(trials, f)

		print("FINISHED, best:")
		pp.pprint(best)
```

**Comment by david-waterworth** [comment]
Maybe implement a callback features, pass an array of callback functions to fmin(), each callback is called before the objective is evaluated passing the trials object. If any callback returns false then stop the search? That way it's easy to extend with early stopping (time based or some other strategy), checkpoint saving (pickle the trials object), write tensorboard metrics etc.

**Comment by CameronFraser** [comment]
The Optuna approach is nice https://optuna.readthedocs.io/en/stable/tutorial/pruning.html

**Comment by phuhavan** [comment]
> @jasonminsookim something like this. it processes trials in chunks, args.trials can be a filename (serial execution) or a mongo-url (parallel execution):
> 
> ```
>                 if args.trials.startswith("mongo://"):
> 			trials = MongoTrials(args.trials, exp_key=args.exp_key)
> 			mongomode = True
> 		else:
> 			mongomode = False
> 			try:
> 				print("Trying to load trials from '{}' ...".format(args.trials))
> 				trials = pickle.load(open(args.trials, "rb"))
> 				print("Found saved Trials! Loading...")
> 				trials_step = 1
> 				print("Rerunning: {} finished trials, will run {} more trials".format(
> 					len(trials.trials), args.max_trials))
> 			except:  # create a new trials object and start searching
> 				print("Starting new trials:", args.trials)
> 				trials = Trials()
> 
> 		for i in range(args.max_trials):
> 			fmin(
> 				fn=optimizer.f,
> 				space=ns.space,
> 				algo=tpe.suggest,
> 				trials=trials,
> 				max_evals=len(trials) + 64, # this limits the amount of (mongo)workers doing things in parallel
> 				max_queue_len=32,
> 				verbose=args.verbose
> 			)
> 
> 			print("Currently finished:",len(trials))
> 			print("Average best loss:",trials.average_best_error())
> 			print("Current best loss:",trials.best_trial['result']['loss'])
> 
> 	        	best = trials.best_trial['result']['space']
> 			pp.pprint(best)
> 	
>                         # check for abort crit somewhere around here
> 
> 			filename = 'models/params_'+ str(trials.best_trial['result']['loss']) + "_" + \
> 				datetime.datetime.now().strftime("%Y-%m-%d_%H:%M:%S") + '.pickle'
> 			with open(filename, 'wb') as output:
> 				pickle.dump(best,
> 							output, pickle.HIGHEST_PROTOCOL)
> 			print('saved as', filename)
> 			# save the trials object, maybe
> 			if not mongomode:
> 				with open(args.trials, "wb") as f:
> 					pickle.dump(trials, f)
> 
> 		print("FINISHED, best:")
> 		pp.pprint(best)
> ```

Thank you for your great idea. However, I am wondering whether the model built by TPE is reused for the next iteration and how can we be sure the model is reused?

**Comment by QuentinAmbard** [comment]
I've created a PR for this feature: https://github.com/hyperopt/hyperopt/pull/661

Feedback is welcome!

**Comment by ctselas** [comment]
@maxpumperla this would be a useful functionality! And probably I will use it as soon as it is released (thanks @QuentinAmbard )

But I came here because I was looking for something else related to early stopping.  I am using hyperopt for XGBoost with early stopping inside the XGBoost. I was wondering if there is a way to retrieve this value when it happens from trials. 

Thank you in advance and sorry that my question is irrelevant with the topic.

**Comment by srowen** [comment]
(I think this can be closed as #661 was merged)

**Comment by vperiyasamy** [comment]
@QuentinAmbard does your implementation handle parallel trials? I made a comment on the PR but it seems like the logic `trials.trials[len(trials.trials)-1]` doesn't actually refer to the most recently completed trial when they're executed in parallel. I guess in this case I should filter for completed jobs and sort by `refresh_time`?

**Comment by QuentinAmbard** [comment]
@vperiyasamy good point, this has not been tested for parallel trials. In this case should we use refresh_time & also filter on status ?
